### PR TITLE
RCD Update (GUI, Smart loading, airlock configs)

### PR DESCRIPTION
**Pull request by Gandalf2k15**
Finally completed it.
This is a port/recreation from Paradise. This update includes many cool changes.

**Changes:**
- RCD now has a GUI instead of craptext.
- RCD is now able to modify airlocks via GUI.
- RCD creates proper windows.
- ERT's loaded RCD is now called a combat RCD.
- RCD's CMC's are now smart loadable. (e.g. if you have 96 units it will now take 4 from cartridge instead of being retarded.
- RCD CMC's come in normal: 50U and large: 100U.
- RCD's now use smart construction for windows.
- RCD CMC's are now smart cartridges, they will fill by 10 for normal or 30 for large, meaning no more crap when loading.

**GUI Images:**
Main Menu:
image
Airlock Settings:
image

**Demo**
https://dl.dropboxusercontent.com/u/109295987/ShareX/2016/07/2016-07-16_02-08-58.mp4
Ignore any glitches in this video, they were fixed.

This will make using an RCD much less cancerous, its a must have.

### CODE REVIEW

**Comment by dreamcrab** [comment]
looks good to me i guess +7

**Comment by ghost** [comment]
This is really nifty +1.

**Comment by GenericGoose** [comment]
hek yeah
i'd suggest changing word 'ammo' to 'charges' or something i dunno

**Comment by GenericGoose** [comment]
can it make windows that face different directions?

**Comment by Gandalf2k15** [comment]
Yeah it makes them smartly, eg finds a window thenmakes it join up

**Comment by Spacedong** [comment]
holy fug yes, btw 2 heads approve pls merge

**Comment by Badmex** [comment]
Nice

**Comment by Voratius** [comment]
being able to choose what type of airlock you make is pretty neato, +1/3rd

**Comment by GeneralGoose** [comment]
gas the konflikts merge war now

**Comment by GenericGoose** [comment]
gibe merge asap tag

**Comment by ghost** [comment]
> Needs code review

R E A D

**Comment by GenericGoose** [comment]
why it has ready tag lel

**Comment by ghost** [comment]
what ready tag???

**Comment by Pyko1** [comment]
Prolly the ready tag that is used to sort these PR's and issues on waffle.io that you just removed.

**Comment by ghost** [comment]
fuck

**Comment by Nebyab** [comment]
i like the old ui
this seems fine though

**Comment by Nebyab** [comment]
to clarify, the ui comment was a joke

gandalf had a little autistic freakout, so just thought i'd be clear

that better gandalf?

**Comment by Gandalf2k15** [comment]
Okay only needs code review now pls

**Comment by JohnGinnane** [comment]
Jesus Christ this is beautiful

**Comment by JohnGinnane** [comment]
I have one minor suggestion. Is it possible to have the selected airlock button be green like the RCD mode? Makes it much easier to see than "Selected airlock: engineering"

File: code/game/objects/items/weapons/RCD.dm [hunk 1]
```diff
@@ -11,312 +11,180 @@ RCD
 	icon_state = "rcd"
 	opacity = 0
 	density = 0
-	anchored = 0
+	anchored = 0.0
```
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm]: Why. Is this something from another codebase?

Comment (TDcoolguy300) [inline code/game/objects/items/weapons/RCD.dm]: It's from Paradise

Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm]: I'm not sure we need to say anchored is a decimal, unless it's used like this elsewhere

Comment (TDcoolguy300) [inline code/game/objects/items/weapons/RCD.dm]: No we don't, but coders in the past wrote the 0 as 0.0, it's still the same to the game when it's checking for it.

**Comment by Gandalf2k15** [comment]
I tried and it never worked, I couldnt do it.

File: code/game/objects/items/weapons/RCD.dm [hunk 2]
```diff
@@ -11,312 +11,180 @@ RCD
...
+	force = 10.0
+	throwforce = 10.0
 	throw_speed = 3
 	throw_range = 5
 	w_class = 3
 	materials = list(MAT_METAL=100000)
 	origin_tech = "engineering=4;materials=2"
-	req_access_txt = "11"
-	var/datum/effect_system/spark_spread/spark_system
+	var/max_matter = 100
```
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm]: This is a change in the max matter. You've reduced it from 160. Is this intentional?

File: code/game/objects/items/weapons/RCD.dm (hunk 1 above)
Comment (Carbonhell) [inline code/game/objects/items/weapons/RCD.dm]: we don't allow .0, cause it's plain useless, just put the normal numbers

File: code/game/objects/items/weapons/RCD.dm [hunk 3]
```diff
@@ -11,312 +11,180 @@ RCD
...
+		/obj/machinery/door/airlock/maintenance = "Maintenance", /obj/machinery/door/airlock/external = "External",
+		/obj/machinery/door/airlock/glass = "Standard (Glass)", /obj/machinery/door/airlock/freezer = "Freezer",
+		/obj/machinery/door/airlock/glass_command = "Command (Glass)", /obj/machinery/door/airlock/glass_engineering = "Engineering (Glass)",
+		/obj/machinery/door/airlock/glass_security = "Security (Glass)", /obj/machinery/door/airlock/glass_medical = "Medical (Glass)",
+		/obj/machinery/door/airlock/mining = "Mining", /obj/machinery/door/airlock/atmos = "Atmospherics",
+		/obj/machinery/door/airlock/research = "Research", /obj/machinery/door/airlock/glass_research = "Research (Glass)",
+		/obj/machinery/door/airlock/glass_mining = "Mining (Glass)", /obj/machinery/door/airlock/glass_atmos = "Atmospherics (Glass)")
 
 /obj/item/weapon/rcd/New()
-	..()
```
Comment (Carbonhell) [inline code/game/objects/items/weapons/RCD.dm:233]: why the hell did ya remove the ..()

File: code/game/objects/items/weapons/RCD.dm [hunk 4]
```diff
@@ -11,312 +11,180 @@ RCD
...
-					if("Research")
-						airlock_type = /obj/machinery/door/airlock/glass_research
-					if("Mining")
-						airlock_type = /obj/machinery/door/airlock/glass_mining
-			else
-				airlock_type = /obj/machinery/door/airlock/glass
-		else
-			airlock_type = /obj/machinery/door/airlock
-
+	var/static/list/allowed_door_types = list(/obj/machinery/door/airlock = "Standard",
```
Comment (Carbonhell) [inline code/game/objects/items/weapons/RCD.dm]: is it necessary for this to be static

File: code/game/objects/items/weapons/RCD.dm [hunk 5]
```diff
@@ -11,312 +11,180 @@ RCD
...
-
 /obj/item/weapon/rcd/Destroy()
-	qdel(spark_system)
-	spark_system = null
 	rcd_list -= src
 	return ..()
 
 /obj/item/weapon/rcd/attackby(obj/item/weapon/W, mob/user, params)
 	..()
-	if(isrobot(user))	//Make sure cyborgs can't load their RCDs
```
Comment (Carbonhell) [inline code/game/objects/items/weapons/RCD.dm:250]: why did you remove this safety check

File: code/game/objects/items/weapons/RCD.dm (hunk 3 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm:233]: This is para code. idk

File: code/game/objects/items/weapons/RCD.dm (hunk 4 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm]: I dunno why the list of available doors would change mid game. Considering a restart would be required.

File: code/game/objects/items/weapons/RCD.dm (hunk 5 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm:250]: Robots dont have hands

File: code/game/objects/items/weapons/RCD.dm (hunk 3 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm:233]: Fixed.

File: code/game/objects/items/weapons/RCD.dm [hunk 6]
```diff
@@ -14,309 +14,178 @@ RCD
...
-		loaded = 1
-	else if(istype(W, /obj/item/stack/sheet/metal) || istype(W, /obj/item/stack/sheet/glass))
-		loaded = loadwithsheets(W, sheetmultiplier, user)
-	else if(istype(W, /obj/item/stack/sheet/plasteel))
-		loaded = loadwithsheets(W, plasteelmultiplier*sheetmultiplier, user) //Plasteel is worth 3 times more than glass or metal
-	if(loaded)
-		user << "<span class='notice'>The RCD now holds [matter]/[max_matter] matter-units.</span>"
-		desc = "A RCD. It currently holds [matter]/[max_matter] matter-units."
-	return
+		if(R.ammoamt < loadamt && max_matter - matter >= R.ammoamt)
```
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm:70]: Any instances of:

```
if(R.ammoamt == 0)
    qdel(W)
```

are just duplicate when you check it later on on like 87

File: code/game/objects/items/weapons/RCD.dm [hunk 7]
```diff
@@ -14,309 +14,178 @@ RCD
...
+			loadamt = R.ammoamt
+			R.ammoamt -= R.ammoamt
+			qdel(W)
+		else if(max_matter - matter < loadamt)
+			loadamt = max_matter - matter
+			R.ammoamt -= max_matter - matter
+			R.desc = "Highly compressed matter for the RCD. It currently has [R.ammoamt] units left."
+			if(R.ammoamt == 0)
```
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm]: Can you maybe check if <= as it's safer, even if unlikely

File: code/game/objects/items/weapons/RCD.dm [hunk 8]
```diff
@@ -14,309 +14,178 @@ RCD
...
+				qdel(W)
+		else
+			R.ammoamt -= loadamt
+			R.desc = "Highly compressed matter for the RCD. It currently has [R.ammoamt] units left."
+			if(R.ammoamt == 0)
+				qdel(W)
+		if(R.ammoamt == 0)
```
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm]: Same here

File: code/game/objects/items/weapons/RCD.dm [hunk 9]
```diff
@@ -325,211 +194,205 @@ RCD
 	switch(mode)
 		if(1)
 			if(istype(A, /turf/space))
-				var/turf/space/S = A
-				if(useResource(floorcost, user))
-					user << "<span class='notice'>You start building floor...</span>"
+				if(useResource(1, user))
```
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm:191]: Original floor cost was 2 resources

File: code/game/objects/items/weapons/RCD.dm [hunk 10]
```diff
@@ -325,211 +194,205 @@ RCD
...
 				return 0
 
 			if(istype(A, /turf/simulated/floor))
-				var/turf/simulated/floor/F = A
-				if(checkResource(wallcost, user))
-					user << "<span class='notice'>You start building wall...</span>"
-					playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
-					if(do_after(user, walldelay, target = A))
-						if(!useResource(wallcost, user)) return 0
+				if(checkResource(3, user))
```
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm:200]: Original wall cost was 16

File: code/game/objects/items/weapons/RCD.dm [hunk 11]
```diff
@@ -325,211 +194,205 @@ RCD
...
-								qdel(T)
-								useResource(-airlockcost, user)
-								return 0
-							T.autoclose = 1
-							return 1
-						return 0
-					else
-						user << "<span class='warning'>There is another door here!</span>"
-						return 0
+				if(checkResource(10, user))
```
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm:213]: Same with airlocks; 16

File: code/game/objects/items/weapons/RCD.dm [hunk 12]
```diff
@@ -325,211 +194,205 @@ RCD
...
-				var/turf/simulated/wall/W = A
-				if(istype(W, /turf/simulated/wall/r_wall) && !canRturf)
+				if(istype(A, /turf/simulated/wall/r_wall) && !canRwall)
 					return 0
-				if(checkResource(deconwallcost, user))
-					user << "<span class='notice'>You start deconstructing wall...</span>"
-					playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
-					if(do_after(user, deconwalldelay, target = A))
-						if(!useResource(deconwallcost, user)) return 0
+				if(checkResource(5, user))
```
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm:233]: Deconstructing walls were 26

File: code/game/objects/items/weapons/RCD.dm [hunk 13]
```diff
@@ -325,211 +194,205 @@ RCD
...
-					return 0
-				if(istype(F, F.baseturf))
-					user << "<span class='notice'>You can't dig any deeper!</span>"
-					return 0
-				else if(checkResource(deconfloorcost, user))
-					user << "<span class='notice'>You start deconstructing floor...</span>"
-					playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
-					if(do_after(user, deconfloordelay, target = A))
-						if(!useResource(deconfloorcost, user)) return 0
+				if(checkResource(5, user))
```
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm:245]: Floors were 33

File: code/game/objects/items/weapons/RCD.dm [hunk 14]
```diff
@@ -325,211 +194,205 @@ RCD
...
 						return 1
 				return 0
 
 			if(istype(A, /obj/machinery/door/airlock))
-				if(checkResource(deconairlockcost, user))
-					user << "<span class='notice'>You start deconstructing airlock...</span>"
-					playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
-					if(do_after(user, deconairlockdelay, target = A))
-						if(!useResource(deconairlockcost, user)) return 0
+				if(checkResource(20, user))
```
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm:257]: And airlocks were only 16

File: code/game/objects/items/weapons/RCD.dm [hunk 15]
```diff
@@ -325,211 +194,205 @@ RCD
...
 			if(istype(A, /obj/structure/grille))
-				var/obj/structure/grille/G = A
-				if(!G.shock(user, 90)) //if it's shocked, try to shock them
-					if(useResource(decongrillecost, user))
-						user << "<span class='notice'>You start deconstructing the grille...</span>"
-						activate()
-						playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
-						qdel(A)
-						return 1
+				if(!checkResource(2, user))
```
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm:271]: And windows were another 8 on top of that

**Comment by JohnGinnane** [comment]
There a lot of costs changed because their RCD max value was lower than ours

File: code/game/objects/items/weapons/RCD.dm [hunk 16]
```diff
@@ -14,309 +14,178 @@ RCD
...
-	src.spark_system = new /datum/effect_system/spark_spread
-	spark_system.set_up(5, 0, src)
-	spark_system.attach(src)
 	rcd_list += src
+	door_accesses_list = list()
+	for(var/access in get_all_accesses())
+		door_accesses_list[++door_accesses_list.len] = list("name" = get_access_desc(access), "id" = access, "enabled" = (access in door_accesses))
 	return
-
+	..()
```
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm]: This will never happen as the return right above it will finish proc

File: code/game/objects/items/weapons/RCD.dm [hunk 17]
```diff
@@ -14,309 +14,178 @@ RCD
...
-		return
-	var/loaded = 0
 	if(istype(W, /obj/item/weapon/rcd_ammo))
+		var/loadamt = 10
+		if(istype(W, /obj/item/weapon/rcd_ammo/large))
+			loadamt = 30 //makes lyfe easier
 		var/obj/item/weapon/rcd_ammo/R = W
-		if((matter + R.ammoamt) > max_matter)
-			user << "<span class='warning'>The RCD can't hold any more matter-units!</span>"
+		if(matter >= max_matter)
```
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm:66]: Should we also cap matter to max matter here?

File: code/game/objects/items/weapons/RCD.dm (hunk 7 above)
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm]: In fact remove this instance of the remove and put <= down to the next one

File: code/game/objects/items/weapons/RCD.dm [hunk 18]
```diff
@@ -14,309 +14,178 @@ RCD
...
+		else if(max_matter - matter < loadamt)
+			loadamt = max_matter - matter
+			R.ammoamt -= max_matter - matter
+			R.desc = "Highly compressed matter for the RCD. It currently has [R.ammoamt] units left."
+			if(R.ammoamt == 0)
+				qdel(W)
+		else
+			R.ammoamt -= loadamt
+			R.desc = "Highly compressed matter for the RCD. It currently has [R.ammoamt] units left."
+			if(R.ammoamt == 0)
```
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm]: Remove this, also a duplicate

File: code/game/objects/items/weapons/RCD.dm [hunk 19]
```diff
@@ -14,309 +14,178 @@ RCD
...
+		. = 1
 
 /obj/item/weapon/rcd/proc/activate()
-	playsound(src.loc, 'sound/items/Deconstruct.ogg', 50, 1)
+	playsound(loc, 'sound/items/Deconstruct.ogg', 50, 1)
 
 
 /obj/item/weapon/rcd/afterattack(atom/A, mob/user, proximity)
-	if(!proximity) return 0
+	if(!proximity) return
```
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm]: Might want to add the 0 back in, something could rely on it

**Comment by JohnGinnane** [comment]
I would also suggest checking the delay on constructing and deconstructing things as ours might differ from paradise

File: code/game/objects/items/weapons/RCD.dm [hunk 20]
```diff
@@ -325,211 +194,205 @@ RCD
...
 
-/obj/item/weapon/rcd/proc/checkResource(amount, mob/user)
+/obj/item/weapon/rcd/proc/checkResource(var/amount, var/mob/user)
 	return matter >= amount
-/obj/item/weapon/rcd/borg/useResource(amount, mob/user)
+/obj/item/weapon/rcd/borg/useResource(var/amount, var/mob/user)
 	if(!isrobot(user))
 		return 0
-	return user:cell:use(amount * 72) //borgs get 1.3x the use of their RCDs
+	return user:cell:use(amount * 160)
```
Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm:355]: Interesting syntax...

Comment (JohnGinnane) [inline code/game/objects/items/weapons/RCD.dm:355]: Ah it's a shortcut

File: code/game/objects/items/weapons/RCD.dm (hunk 6 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm:70]: Fixing.

Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm:70]: Infact it was fixed.

File: code/game/objects/items/weapons/RCD.dm (hunk 7 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm]: Doooone

File: code/game/objects/items/weapons/RCD.dm (hunk 8 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm]: Dooone

File: code/game/objects/items/weapons/RCD.dm (hunk 9 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm:191]: Intended

File: code/game/objects/items/weapons/RCD.dm (hunk 10 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm:200]: Intended

File: code/game/objects/items/weapons/RCD.dm (hunk 11 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm:213]: intended

File: code/game/objects/items/weapons/RCD.dm (hunk 12 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm:233]: Intended

File: code/game/objects/items/weapons/RCD.dm (hunk 13 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm:245]: Intended

File: code/game/objects/items/weapons/RCD.dm (hunk 14 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm:257]: Intended

File: code/game/objects/items/weapons/RCD.dm (hunk 15 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm:271]: Intended

File: code/game/objects/items/weapons/RCD.dm (hunk 17 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm:66]: Aye, always good to have safety checks.

File: code/game/objects/items/weapons/RCD.dm (hunk 16 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm]: Fixed

File: code/game/objects/items/weapons/RCD.dm (hunk 17 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm:66]: Fixed

File: code/game/objects/items/weapons/RCD.dm (hunk 18 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm]: Fixed

File: code/game/objects/items/weapons/RCD.dm (hunk 19 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm]: Fixed.

File: code/game/objects/items/weapons/RCD.dm (hunk 20 above)
Comment (Gandalf2k15) [inline code/game/objects/items/weapons/RCD.dm:355]: Meep

**Comment by Gandalf2k15** [comment]
Uploading fixes, all resource/time things were intended.

**Comment by JohnGinnane** [comment]
Everything looks good. Merging
